builtin/credential/aws/path_role.go

```diff
@@ -319,6 +319,132 @@ func (b *backend) setRole(ctx context.Context, s logical.Storage, roleName strin
 	return nil
 }
 
+// initialize is used to initialize the AWS roles
+func (b *backend) initialize(ctx context.Context, req *logical.InitializationRequest) error {
+
+	// on standbys and DR secondaries we do not want to run any kind of upgrade logic
+	if b.System().ReplicationState().HasState(consts.ReplicationPerformanceStandby | consts.ReplicationDRSecondary) {
+		return nil
+	}
+
+	// Initialize only if we are either:
+	//   (1) A local mount.
+	//   (2) Are _NOT_ a replicated performance secondary
+	if b.System().LocalMount() || !b.System().ReplicationState().HasState(consts.ReplicationPerformanceSecondary) {
+
+		s := req.Storage
+
+		logger := b.Logger().Named("initialize")
+		logger.Info("starting initialization")
+
+		go func() {
+			// The vault will become unsealed while this goroutine is running,
+			// so we could see some role requests block until the lock is
+			// released.  However we'd rather see those requests block (and
+			// potentially start timing out) than allow a non-upgraded role to
+			// be fetched.
+			b.roleMutex.Lock()
+			defer b.roleMutex.Unlock()
+
+			upgraded, err := b.upgrade(ctx, s)
+			if err != nil {
+				logger.Error("error running initialization", "error", err)
+				return
+			}
+			if upgraded {
+				logger.Info("an upgrade was performed during initialization")
+			}
+		}()
+
+	}
+
+	return nil
+}
+
+// awsVersion stores info about the the latest aws version that we have
+// upgraded to.
+type awsVersion struct {
+	RoleVersion int `json:"role_verison"`
+}
+
+// upgrade does an upgrade, if necessary
+func (b *backend) upgrade(ctx context.Context, s logical.Storage) (bool, error) {
+
+	entry, err := s.Get(ctx, "config/version")
+	if err != nil {
+		return false, err
+	}
+
+	// if there is no persisted version, we need to upgrade
+	if entry == nil {
+		err = b.upgradeRoles(ctx, s)
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+
+	var version awsVersion
+	err = entry.DecodeJSON(&version)
+	if err != nil {
+		return false, err
+	}
+
+	// upgrade if persisted roleVersion is out of date
+	switch version.RoleVersion {
+	case 0:
+		fallthrough
+	case 1:
+		fallthrough
+	case 2:
+		err = b.upgradeRoles(ctx, s)
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+
+	case currentRoleStorageVersion:
+		return false, nil
+
+	}
+	return false, fmt.Errorf("unrecognized role version: %d", version.RoleVersion)
+}
+
+// upgradeRoles upgrades the various aws roles
+func (b *backend) upgradeRoles(ctx context.Context, s logical.Storage) error {
```

Comment:
The way it is now, if you don't call upgradeRoles, you don't actually persist the new version. Combine this into the function above; the role specific stuff can go into the switch case, and then after the switch case is run update to the latest version and persist.